Add the v1 content for xp-arch-framework (#177)
upbound/docs: content/xp-arch-framework/interface-integrations/platform-continuity.md

---
title: "Platform Continuity"
weight: 9
description: "A guide for how to integrate control planes with DR tools"
---

Platform continuity is a topic that encompasses everything you need to do to keep your platform up and running. This content is currently focused on providing guidance for how to perform disaster recovery with Crossplane.

## Disaster Recovery

> [Velero](https://velero.io/) is a popular open source solution in the Kubernetes ecosystem to safel perform backup and restore operations against Kubernetes cluster resources. We recommend using Velero as the key ingredient for how to do disaster recovery with your Crossplane control planes. We will cover best practices for using Velero in the context of Crossplane.

Reviewer: ```suggestion
[Velero](https://velero.io/) is a popular open source solution in the Kubernetes ecosystem to safely perform backup and restore operations against Kubernetes cluster resources. We recommend using Velero in any disaster recovery plan for your Crossplane control planes. We will cover best practices for using Velero in the context of Crossplane.
```

### In the context of Crossplane

It's important to remember:

1. Crossplane is concerned with ensuring the configuration of your resource. 
2. Crossplane **isn't concerned** with the _internal_ state of your resources. 

Crossplane is concerned with enforcing two things:

1. the resources you declared exist.
2. the configuration of those resources matches what you declared. If it doesn't, it will attempt to reconcile to the desired state.

Crossplane **isn't** concerned with the state of what's happening _within_ the resource. For example, consider a CloudSQL instance created with Crossplane using the following manifest:

```yaml
apiVersion: sql.gcp.upbound.io/v1beta1
kind: DatabaseInstance
spec:
  forProvider:
    databaseVersion: MYSQL_5_7
    region: us-central1
    settings:
      - diskSize: 20
        tier: db-f1-micro
```

> Crossplane will ensure whatever instance gets created in GCP has a disk size of 20 GB and that it uses MySQL version 5.7; it will continuously ensure the configuration of the instance matches these parameters. However, Crossplane has no involvement with what goes on _inside_ the CloudSQL instance. If someone was using this database to store records of users for a line-of-business app, Crossplane is oblivious.

Reviewer: "Disaster Recovery for Crossplane != disaster recovery for your business critical state, like database or cloud storage data."

Likewise, if your control plane fails and goes offline, that doesn't necessarily mean the resources under management by that control plane will be affected--it means the configuration of those resources won't be reconciled while the control plane is offline, so configuration drift could occur. Using the database example above, if the control plane managing the CloudSQL instance fails, that doesn't necessarily mean your database is now offline. 

### What State to Capture vs Exclude

> You should configure Velero to capture state that's only relevant for Crossplane. THe following is a sample Velero backup object that illustrates which Kubernetes resources are applicable to be backed up and which ones can be safely excluded:

Reviewer: ```suggestion
You should configure Velero to capture state that's only relevant for Crossplane. The following is a sample Velero backup object that illustrates which Kubernetes resources are applicable to be backed up and which ones can be safely excluded:
```

{{< /hint >}}

### Backup Schedules

We recommend establishing a Velero schedule to periodically take snapshots for each control plane. THe following is a sample Velero schedule object that illustrates a recommended schedule:

```yaml
apiVersion: velero.io/v1
kind: Schedule
metadata:
  name: daily-backup
  namespace: velero
spec:
  schedule: '@every 24h'
  template:
    includedNamespaces:
    - '*'
    excludedNamespaces:
    - kube-node-lease
    - kube-system
    - kube-public
    - velero
    excludedResources:
    - bindings
    - componentstatuses
    - endpoints
    - events
    - limitranges
    - nodes
    - persistentvolumeclaims
    - persistentvolume
    - pods
    - podtemplates
    - replicationcontrollers
    - resourcequotas
    - services
    - daemonsets
    - deployments
    - replicasets
    - statefulsets
    - cronjobs
    - jobs
    includeClusterResources: true
    snapshotVolumes: false
    metadata: {}
    ttl: 168h0m0s
```

For each control plane that you create, create three Velero `Schedule` objects illustrated above with the following schedules:

- Hourly
- Daily
- Weekly

> We think this is a good baseline, but you can customize the frequency of the backups according to the Recovery Point Objective (RPO) for your platform. If you need to be able to restore state more frequently, you will want to make the snapshots more frequent than hourly.

Reviewer: ```suggestion
We think this is a good baseline, but you can customize the frequency of the backups according to the Recovery Point Objective (RPO) for your platform. If you need to be able to restore state more frequently, you will want to snapshot more frequently.
```

### Restore Control Plane from state

To restore backup state into a new cluster:

> 1. Ensure that Crossplane and providers aren't running in a new cluster. Otherwise, the ordering between managed resources, composites, and claims becomes critical. Failure to do so may result in race conditions and, for example, duplication of managed resources.

Reviewer: > Ensure that Crossplane and providers aren't running in a new cluster.

If my backup contains my provider definitions, how do I control their scaling? Is there a field I need to override in my backups or during my restore?

Reviewer 2: `pods` and `deployments` added to  `excludedResources` in velero backup definition - so you only need to check that your new target cluster is not running crossplane or any providers - scale to 0 or add replicas 0 in `controllerConfig` - otherwise folks need to check the backup manifests before restore
